### PR TITLE
#701 Feature Support SSH Keys

### PR DESCRIPTION
**Pull request by jaapmarcus**
As discussed in #701, #700 it would be a great add-on and could be improving the security silghtly.

On my server I have disabled by default the use of login via password. When creating a new account (Internal use) I had every time manually add them the authorised_keys of the user manually

By default all commands are executed within the Hestia as sudo therefore almost no additional risk is taken.

v-add-user-ssh-key
Bash script will check if authorized_keys  exists if not try to create it and then append it
v-list-user-ssh-ssh-key
List all keys in authorized_keys
v-delete-user-ssh-key
Delete selected key form file with use of regex

There is no option for editing the SSH key just remove them and add a new one

Thing to be improved:
* Better way to validate check format of the SSH key.
*Changing to some names?
*Currently the page is only available with https://192.168.0.27:8083/list/key Or an option  needs to be added on index page or preffable user edit page.
https://192.168.0.27:8083/edit/user/?user=test
* May be add option the select multiple keys and delete them at once. 
* Add bash check for valid format SSH key? Don't know if it is possible? 

By default the keys are saved by:
ssh-rsa 64bitencodedstring== random@computer.local

The **random@computer.local** is used for deleting later on there for it must be unique. Option would be to replace it with a random string or even een user generated string however this has also some disadvantages

### CODE REVIEW

**Comment by ScIT-Raphael** [comment]
Awesome, thanks for your contribution! We will review and implement your code as soon as possible!

**Comment by jaapmarcus** [comment]
Added check for valid SSH key in bash script

``TEMP=$(mktemp)
echo "$key" >> "$TEMP"
ssh-keygen -l -f "$TEMP"
if [ ! $? -eq 0 ]; then
	rm "$TEMP"
	exit
fi
rm "$TEMP"``

When it returns true it will add the key to the authorized_keys if not it will fail

File: web/add/key/index.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	if (empty($_POST['v_key'])) $errors[] = __('ssh key');
+    
+        if (!empty($errors[0])) {
+        foreach ($errors as $i => $error) {
+            if ( $i == 0 ) {
+                $error_msg = $error;
+            } else {
+                $error_msg = $error_msg.", ".$error;
+            }
+        }
```
Comment (Lupul) [inline web/add/key/index.php]: what is the purpose of this block?

File: web/add/key/index.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  	$keylist = array();
+  	foreach($data as $key => $value){
+	  	$idlist[] = trim($data[$key]['ID']);
+	  	$keylist[] = trim($data[$key]['KEY']);
+	  	
+  	}
+  	
+  	if(!validateKey($_POST['v_key'])){
+	  	$_SESSION['error_msg']  = __('SSH KEY is invalid');
+  	}
```
Comment (Lupul) [inline web/add/key/index.php]: If the key is invalid is there a reason to process the file further?

File: web/add/key/index.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+     unset($output);
+
+     
+	}
+	
+	// Flush field values on success
+    if (empty($_SESSION['error_msg'])) {
+        $_SESSION['ok_msg'] = __('SSH KEY Added');
+		unlink($v_key);
+		unlink($_POST['v_key']);
```
Comment (Lupul) [inline web/add/key/index.php]: unlink deletes a file, i don't see a file being created prev.

File: web/delete/key/index.php [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+if (!empty($_GET['key'])) {
+    $v_key = escapeshellarg(trim($_GET['key']));
+    $v_key = str_replace('/','\\/', $v_key);
+    echo HESTIA_CMD."v-delete-user-ssh-key ".$user." ".$v_key;
+	exec (HESTIA_CMD."v-delete-user-ssh-key ".$user." ".$v_key);
+	check_return_code($return_var,$output);
+}
+
+//var_dump($return_var,$v_key);
```
Comment (Lupul) [inline web/delete/key/index.php]: commented code should be removed

File: web/list/key/index.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+// Main include
+include($_SERVER['DOCUMENT_ROOT']."/inc/main.php");
+
+exec (HESTIA_CMD . "v-list-user-ssh-key ".escapeshellarg($user)." json", $output, $return_var);
+
+
+$data = json_decode(implode('', $output), true);
+
+//$data = array_reverse($data,true);
```
Comment (Lupul) [inline web/list/key/index.php]: commented code should be removed

File: web/templates/admin/list_key.html [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                                <div id="delete_dialog_<?=$i?>" class="confirmation-text-delete hidden" title="<?=__('Confirmation')?>">
+                                  <p class="confirmation"><?=__('DELETE_KEY_CONFIRM',$key)?></p>
+                                </div>
+                              </a>
+                            </div>
+              
+                          </div>
+                        </div>
+                  </div>
+        <div class="clearfix l-unit__stat-col--left wide-7"><b><?=substr($data[$key]['KEY'], 0,50).'.....'.substr($data[$key]['ID'], -10,10);?></b></div>
```
Comment (Lupul) [inline web/templates/admin/list_key.html]: escape all variables that are rendered in the browser eg: htmlspecialchars($str, ENT_QUOTES)

File: web/add/key/index.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	#todo a sanity check would be nice
+	
+	if (empty($_SESSION['error_msg'])) {
+	exec (HESTIA_CMD."v-add-user-ssh-key ".$user." ".$v_key, $output, $return_var);
+    check_return_code($return_var,$output);
+    
+     unset($output);
+
+     
+	}
```
Comment (Lupul) [inline web/add/key/index.php]: Indentaion is inconsistent (not only in this file), please use 4 spaces and drop the unnecessary newlines

**Review by Lupul — CHANGES_REQUESTED** [review]

File: web/add/key/index.php (hunk 2 above)
Comment (jaapmarcus) [reply web/add/key/index.php]: No,  will change it to switch statement. Then break can be used. 
2nd option would be put every thing in a proper class structure but than a better idea is to rewrite the whole code base.

File: web/add/key/index.php (hunk 3 above)
Comment (jaapmarcus) [reply web/add/key/index.php]: True probably stole it from the create database php. And there its used. Will remove it.

File: web/add/key/index.php (hunk 1 above)
Comment (jaapmarcus) [reply web/add/key/index.php]: None, Will alter it

**Comment by jaapmarcus** [comment]
Thanks for the review. Will push the change tomorrow.

One small question is it true that if a template /web/templates/user/  directory doesn't exists the admin template can be used?

**Comment by jaapmarcus** [comment]
Modified all changes as requested

**Comment by ScIT-Raphael** [comment]
Hi Guys

I merged the pull request to our new branch "feature-ssh-keys", so we can test them on our system using the v-update-sys-hestia-git script according to our docs: https://github.com/hestiacp/hestiacp#installing--testing-development-builds

**Comment by jaapmarcus** [comment]
Currently the /edit/user template hasn't been changed yet. Therefore the only way to acces this function is via the flowing path:

https://server.com:8083/list/key

**Comment by ww7** [comment]
Throw invalid error for ssh-ed25519 type of key. Why it's not accepted?

**Comment by jaapmarcus** [comment]
Due to the check here

https://github.com/hestiacp/hestiacp/blob/main/web/add/key/index.php#L10-L39

Maybe we can safely remove the check...

**Comment by jaapmarcus** [comment]
https://github.com/hestiacp/hestiacp/pull/1161

**Comment by ghost** [comment]
@jaapmarcus rsa and ssh-ed25519 are probably the only key types anyone will use. The check may not be necessary.. If your using certificates you are most likely not a total noob. if ssh-ed25519 being verified as well is going to be overly complicated..
Maybe would be to add a api "v-force-add-key" basically a second set of the scripts that don't check and add a checkbox or something to the GUI that allows key adding to use the no check api script instead?

If your using ssh-ed25519 you probably know enough to get it right with out verification.

**Comment by jaapmarcus** [comment]
See remark regarding pull request

There was a double check one for php and was only allowing dsa or rsa keys and a v-add-user-ssh-key that did a check with ssh-keygen

The first check is removed the second check still remains

**Pull request merged**